kernel/src/actions/mod.rs

```diff
@@ -210,8 +209,16 @@ impl Protocol {
                 )
             );
         }
-        let reader_features = reader_features.map(|f| f.into_iter().map(Into::into).collect());
-        let writer_features = writer_features.map(|f| f.into_iter().map(Into::into).collect());
+        let reader_features = reader_features.and_then(|f| {
+            f.into_iter()
+                .map(feature_or_unknown)
+                .collect::<Option<Vec<ReaderFeature>>>()
```

Comment:
do we need type annotations? feels like it _should_ be able to figure out since this is bound to the type of `Protocol.reader_features`